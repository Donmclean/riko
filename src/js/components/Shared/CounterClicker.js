"use strict";
import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';

//IMPORTANT!: Import actions to dispatch
import getActions from '../../actions/_index';

class CounterClicker extends Component {
    constructor(props) {
        super(props);
        //Get all actions
        this.actions = getActions(this.props.dispatch);
    }

    render() {
        const { state } = this.props;
        return (
            <div>
                <h4>Synchronous Actions</h4>

                <button className="inc" onClick={this.actions.counterActionCreators.shouldIncrement}>INCREMENT</button>
                <button className="dec" onClick={this.actions.counterActionCreators.decrement}>DECREMENT</button>

                <h1 className="counter-val">Counter Reducer State: {state.counterReducer.get('value')}</h1>
            </div>
        );
    }
}

CounterClicker.propTypes = {
    state: PropTypes.object.isRequired,
    dispatch: PropTypes.func.isRequired
};

const mapStateToProps = (state) => {
    return {state};
};

export default connect(mapStateToProps)(CounterClicker);