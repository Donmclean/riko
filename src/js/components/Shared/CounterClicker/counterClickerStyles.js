import styled from 'styled-components';

export const Button = styled.button`
  margin: auto;
`;