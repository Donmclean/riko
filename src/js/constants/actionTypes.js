export const SHOULD_INCREMENT = 'SHOULD_INCREMENT';
export const SHOULD_DECREMENT = 'SHOULD_DECREMENT';
export const INCREMENT = 'INCREMENT';
export const DECREMENT = 'DECREMENT';

export const UPDATE_POSTS = 'UPDATE_POSTS';
export const UPDATE_POST_NUMBER = 'UPDATE_POST_NUMBER';