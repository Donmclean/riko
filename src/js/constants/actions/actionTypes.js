export const INCREMENT = 'INCREMENT';
export const DECREMENT = 'DECREMENT';

export const GET_POSTS = 'GET_POSTS';
export const UPDATE_INPUT_VALUE = 'UPDATE_INPUT_VALUE';