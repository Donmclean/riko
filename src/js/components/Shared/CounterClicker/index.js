"use strict";
import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';

import { Button } from './counterClickerStyles';

//IMPORTANT!: Import actions to dispatch
import getActions from '../../../_ActionCreators';

class CounterClicker extends Component {
    constructor(props) {
        super(props);
        //Get all actions
        this.actions = getActions(this.props.dispatch);
    }

    render() {
        const { state } = this.props;
        return (
            <div>
                <h4>Synchronous Actions</h4>

                <Button onClick={this.actions.counterActionCreators.shouldIncrement}>INCREMENT</Button>
                <Button onClick={this.actions.counterActionCreators.decrement}>DECREMENT</Button>

                <h1>Counter Reducer State: {state.counterReducer.get('value')}</h1>
            </div>
        );
    }
}

CounterClicker.propTypes = {
    state: PropTypes.object.isRequired,
    dispatch: PropTypes.func.isRequired
};

const mapStateToProps = (state) => ({state});

export default connect(mapStateToProps)(CounterClicker);