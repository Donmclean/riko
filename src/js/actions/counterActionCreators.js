import * as types from '../../js/constants/actions/actionTypes';

export const shouldIncrement = () => ({
    type: types.SHOULD_INCREMENT
});

export const increment = () => ({
    type: types.INCREMENT
});

export const shouldDecrement = () => ({
    type: types.SHOULD_DECREMENT
});

export const decrement = () => ({
    type: types.DECREMENT
});