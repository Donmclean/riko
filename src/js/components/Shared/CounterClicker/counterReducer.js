"use strict";
import { Map } from 'immutable';
import * as types from '../../../constants/actionTypes';

//Every reducer should have an initial state variable here
export const initialState = Map({ // Export it for testing purposes
    value: 0
});

const counterReducer = (state = initialState, action) => {
    switch (action.type) {
        case types.SHOULD_INCREMENT: {
            break;
        }
        case types.INCREMENT: {
            return state.update('value', (value) => value + 1);
        }
        case types.SHOULD_DECREMENT: {
            break;
        }
        case types.DECREMENT: {
            return state.update('value', (value) => value - 1);
        }
        default: {
            break;
        }
    }

    return state;
};

export default counterReducer;