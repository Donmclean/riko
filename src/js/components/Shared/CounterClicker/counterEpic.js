import * as types from '../../../constants/actions/actionTypes';

//The *KEY* is to receive an action that will trigger another action that actually has direct influence on state
//We can call this an Epic listener. (eg: 'SHOULD_INCREMENT')
//notice the convention of adding 'SHOULD_' in front of the action type.
export const counterEpic = action$ =>
    action$.ofType(types.SHOULD_INCREMENT)
        .debounceTime(1000)
        .mapTo({ type: types.INCREMENT });
