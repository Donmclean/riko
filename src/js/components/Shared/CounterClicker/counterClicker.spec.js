import React from 'react';
import * as counterActions from './counterActionCreators';
import * as types from '../../../constants/actionTypes';
import counterReducer, { initialState } from './counterReducer';

//Include Redux Observable For 'Epic' Creation
import { epics } from '../../../_Epics';
import { createEpicMiddleware } from 'redux-observable';
const epicMiddleware = createEpicMiddleware(epics);

import { Map } from 'immutable';

//For testing components
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
const middlewares = [epicMiddleware]; //eg: [thunk] for async | eg: [epicMiddleware] for testing epics
const mockStore = configureStore(middlewares);

//Component to be Tested
import CounterClicker from './index';

//Test suite designed to test the functionality of
describe('Counter Test Suite', () => {
    let store;

    beforeEach(() => {
        //Initialize mockStore with initial state of reducer
        store = mockStore({counterReducer: initialState});
    });

    //Action Creator Specific Tests would go here
    describe('Counter Actions', () => {
        it('should create a increment action', () => {
            const expectedAction = {
                type: types.INCREMENT
            };
            expect(counterActions.increment()).toEqual(expectedAction);
        });

        it('should create a decrement action', () => {
            const expectedAction = {
                type: types.DECREMENT
            };
            expect(counterActions.decrement()).toEqual(expectedAction);
        });
    });

    //Reducer Specific Tests would go here
    describe('Counter Reducer', () => {
        it(`${types.INCREMENT} executes successfully`, () => {
            //Setup
            const expectedState = Map({value: 1});

            //Execute
            const newState = counterReducer(initialState, counterActions.increment());

            //Verify
            expect(newState).toEqual(expectedState);
        });

        it(`${types.DECREMENT} executes successfully`, () => {
            //Setup
            const expectedState = Map({value: -1});

            //Execute
            const newState = counterReducer(initialState, counterActions.decrement());

            //Verify
            expect(newState).toEqual(expectedState);
        });
    });

    //Epic Specific Tests would go here
    describe('Counter Epic', () => {
        it(`${types.SHOULD_INCREMENT} executes/returns the expected actions successfully`, (done) => {
            store.dispatch(counterActions.shouldIncrement());

            //Since the counter epic has a debounceTime attached we have to provide
            //a setTimeout after the dispatch function is called for the Epic to complete.
            setTimeout(() => {
                expect(store.getActions()).toMatchSnapshot();
                done();
            }, 1100);

        });
    });

    //Component Specific Tests would go here
    describe('Counter Component', () => {
        it('renders <CounterClicker /> component successfully in to the DOM with required props & state', () => {
            //Verify props
            const wrapper = shallow(<CounterClicker store={store}/>);
            expect(wrapper.prop('store')).toEqual(store);

            //Verify that initial State of the store is rendered correctly
            expect(wrapper.state().storeState).toEqual(store.getState());

            //Render Component Tree
            const component = renderer.create(
                <CounterClicker store={store}/>
            );

            //Verify Component Snapshot
            //call component to JSON before matching snapshot
            expect(component.toJSON()).toMatchSnapshot();
        });
    });
});