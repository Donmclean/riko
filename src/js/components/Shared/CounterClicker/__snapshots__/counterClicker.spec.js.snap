exports[`Counter Test Suite Counter Component renders <CounterClicker /> component successfully in to the DOM with required props & state 1`] = `
<div>
  <h4>
    Synchronous Actions
  </h4>
  <button
    className="HaWdQ"
    onClick={[Function]}>
    INCREMENT
  </button>
  <button
    className="HaWdQ"
    onClick={[Function]}>
    DECREMENT
  </button>
  <h1>
    Counter Reducer State: 
    0
  </h1>
</div>
`;

exports[`Counter Test Suite Counter Epic SHOULD_INCREMENT executes/returns the expected actions successfully 1`] = `
Array [
  Object {
    "type": "SHOULD_INCREMENT",
  },
  Object {
    "type": "INCREMENT",
  },
]
`;
